### ISSUE TITLE
`LEAVE` phaser fires when errorring in argument binding

### ISSUE THREAD

**Issue by dakkar** [Raku/problem-solving#361]
```raku
sub foo(Int:D :$x) {
  ENTER { say "entering foo" };
  LEAVE { say "leaving foo" };
}
foo()
```

produces:

```
leaving foo
Parameter '$x' of routine 'foo' must be an object instance of type
'Int', not a type object of type 'Int'.  Did you forget a '.new'?
```

Two things feel surprising to me:
1. the `LEAVE` block was run, without the `ENTER` block being run
2. the `LEAVE` block was run ever though the sub was not "really" executed (I know that signature binding happens sort-of inside the sub, but I feel that should be an implementation detail, not something visible at the language level)

**Comment by dakkar** [comment]
See also the discussion at https://s.thenautilus.net/notes/99rjsevs52

**Comment by raiph** [comment]
I don't think this issue benefits from mentioning the ENTER phaser, at least not in the title and in comments from here on (unless I'm missing something).

This is first and foremost because I didn't see anyone (in the linked discussion) saying they're surprised by how the ENTER phaser is interacting with signature binding.

But it's also because the design does not keep ENTER and LEAVE phasers symmetrical. For example, the queue of ENTER phasers (there can be more than one) for a block is aborted if there's an exception thrown by an ENTER phaser, whereas the LEAVE phaser queue keeps on trucking even if an exception is thrown by a LEAVE phaser. So expecting symmetry between the ENTER and LEAVE phasers in general is a mistake and it may just be that expecting symmetry in this trial bind aspect is also a mistake.

In summary, I currently suspect mentioning ENTER in any of our subsequent discussion is a red herring, and the mention in the title would best be removed.

**Comment by raiph** [comment]
I don't think this issue benefits from mentioning surprise, at least not in the title.
 
Surprise of this sort is relative to surmise, and both are subjective. I think the main thing that matters is to decide whether things are as they are for good reason. If they are, then perhaps we start a doc issue that covers the surmise/surprise. If they are not, then perhaps we change Raku(do).

(All of us must surmise things to get by, but that doesn't mean any particular surmise is a good one. It may just be something we need to let go, no matter how much we would prefer to hold on to it. Similarly, all of us will be surprised on occasion, but that doesn't mean any particular surprise is bad one. It may just be a learning opportunity, no matter how strongly we think it's a problem.)

So I suggest we try to nail things down to an objective problem/expectation statement without recourse to a subjective surmise/surprise statement (and edit the title to match that statement if/when we get there).

**Comment by raiph** [comment]
Perhaps the reason why things are as they are is in case anyone writes a signature that requires post bind attempt clean up, and LEAVE was considered the appropriate hook.

**Comment by 2colours** [comment]
It seems to me that this comment disappeared, even though it might be good to at least be able to read it:

> I don't think this issue benefits from mentioning the ENTER phaser, at least not in the title and in comments from here on (unless I'm missing something).

> This is first and foremost because I didn't see anyone (in the linked discussion) saying they're surprised by how the ENTER phaser is interacting with signature binding.

> But it's also because the design does not keep ENTER and LEAVE phasers symmetrical. For example, the queue of ENTER phasers (there can be more than one) for a block is aborted if there's an exception thrown by an ENTER phaser, whereas the LEAVE phaser queue keeps on trucking even if an exception is thrown by a LEAVE phaser. So expecting symmetry between the ENTER and LEAVE phasers in general is a mistake and it may just be that expecting symmetry in this trial bind aspect is also a mistake.

> In summary, I currently suspect mentioning ENTER in any of our subsequent discussion is a red herring, and the mention in the title would best be removed.
